File: BACKEND/shared/clerk/clerkClient.js
// Helper goi Clerk Backend API.
// Yeu cau bien moi truong CLERK_SECRET_KEY trong BACKEND/.env.
// Khi user bi khoa, ta goi Clerk de revoke tat ca session cua ho.

const { getEnv } = require("../config/env");

const CLERK_API_BASE = "https://api.clerk.com/v1";

function getSecretKey() {
  const key = getEnv("CLERK_SECRET_KEY", "");
  return key.trim();
}

function isClerkConfigured() {
  return Boolean(getSecretKey());
}

async function clerkFetch(path, options = {}) {
  const secretKey = getSecretKey();
  if (!secretKey) {
    throw new Error("CLERK_SECRET_KEY chua duoc cau hinh trong BACKEND/.env.");
  }

  const response = await fetch(`${CLERK_API_BASE}${path}`, {
    ...options,
    headers: {
      Authorization: `Bearer ${secretKey}`,
      "Content-Type": "application/json",
      ...(options.headers || {}),
    },
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Clerk API ${response.status}: ${text || response.statusText}`);
  }

  if (response.status === 204) return null;

  const contentType = response.headers.get("content-type") || "";
  if (!contentType.includes("application/json")) return null;

  return response.json();
}

async function listSessionsForUser(userId) {
  const data = await clerkFetch(`/sessions?user_id=${encodeURIComponent(userId)}&status=active`);
  return Array.isArray(data) ? data : data?.data || [];
}

async function revokeSession(sessionId) {
  return clerkFetch(`/sessions/${encodeURIComponent(sessionId)}/revoke`, {
    method: "POST",
  });
}

async function revokeAllSessionsForUser(userId) {
  if (!isClerkConfigured()) {
    return { revoked: 0, skipped: true };
  }

  const sessions = await listSessionsForUser(userId);
  let revoked = 0;
  for (const session of sessions) {
    try {
      await revokeSession(session.id);
      revoked += 1;
    } catch (error) {
      console.warn(`Khong revoke duoc session ${session.id}:`, error.message);
    }
  }
  return { revoked, skipped: false };
}

module.exports = {
  isClerkConfigured,
  revokeAllSessionsForUser,
};

File: BACKEND/admin/Services/ClerkClient.cs
using System.Net.Http.Headers;
using System.Text.Json;

namespace MyPuppy.Admin.Services;

public class ClerkClient
{
    private const string ClerkApiBase = "https://api.clerk.com/v1";

    private readonly HttpClient _httpClient;
    private readonly string _secretKey;
    private readonly ILogger<ClerkClient> _logger;

    public ClerkClient(HttpClient httpClient, IConfiguration configuration, ILogger<ClerkClient> logger)
    {
        _httpClient = httpClient;
        _secretKey = configuration["Clerk:SecretKey"] ?? string.Empty;
        _logger = logger;
    }

    public bool IsConfigured => !string.IsNullOrWhiteSpace(_secretKey);

    public async Task<(int Revoked, bool Skipped, string? Error)> RevokeAllSessionsForUserAsync(string userId)
    {
        if (!IsConfigured)
        {
            return (0, true, null);
        }

        try
        {
            var sessions = await ListSessionsAsync(userId);
            int revoked = 0;
            foreach (var sessionId in sessions)
            {
                try
                {
                    await RevokeSessionAsync(sessionId);
                    revoked++;
                }
                catch (Exception ex)
                {
                    _logger.LogWarning(ex, "Khong revoke duoc Clerk session {SessionId}", sessionId);
                }
            }
            return (revoked, false, null);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Loi khi goi Clerk Backend API cho user {UserId}", userId);
            return (0, true, ex.Message);
        }
    }

    private async Task<List<string>> ListSessionsAsync(string userId)
    {
        using var request = new HttpRequestMessage(
            HttpMethod.Get,
            $"{ClerkApiBase}/sessions?user_id={Uri.EscapeDataString(userId)}&status=active");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _secretKey);

        using var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        await using var stream = await response.Content.ReadAsStreamAsync();
        using var doc = await JsonDocument.ParseAsync(stream);

        var root = doc.RootElement;
        var sessionIds = new List<string>();

        // Clerk có thể trả mảng trực tiếp hoặc { data: [...] }
        JsonElement array;
        if (root.ValueKind == JsonValueKind.Array)
        {
            array = root;
        }
        else if (root.TryGetProperty("data", out var data) && data.ValueKind == JsonValueKind.Array)
        {
            array = data;
        }
        else
        {
            return sessionIds;
        }

        foreach (var session in array.EnumerateArray())
        {
            if (session.TryGetProperty("id", out var id) && id.ValueKind == JsonValueKind.String)
            {
                sessionIds.Add(id.GetString()!);
            }
        }
        return sessionIds;
    }

    private async Task RevokeSessionAsync(string sessionId)
    {
        using var request = new HttpRequestMessage(
            HttpMethod.Post,
            $"{ClerkApiBase}/sessions/{Uri.EscapeDataString(sessionId)}/revoke");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _secretKey);

        using var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
    }
}
